hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/execution/traceability/HederaTracer.java

```diff
@@ -133,9 +166,93 @@ private void trackNewAction(final MessageFrame messageFrame, final Consumer<Soli
     }
 
     private void finalizeActionFor(final SolidityAction action, final MessageFrame frame, final State frameState) {
+
+        switch (frameState) {
+            case NOT_STARTED, CODE_EXECUTING, CODE_SUSPENDED:
+                {
+                    // these states are not "final" states needing to finalize the actions
```

Comment:
Good luck testing this.  It's not worth putting in an induced failure, I'm OK with these being expected as non-reachable and hence untested.  With this line in place it will at least put the idea in people heads if they are debugging this is what could be wrong.

Reply:
Yes, it is mainly for readers.  And for `javac` pointing out missing cases (enum values not handled).  Which was what the problem initially was: no handling for the `COMPLETED_FAILED` condition.